README.md
### This project is deprecated. For modern Elm support in VSCode use https://marketplace.visualstudio.com/items?itemName=Elmtooling.elm-ls-vscode (GitHub: https://github.com/elm-tooling/elm-language-client-vscode)

---

# Elm support for Visual Studio Code

![Error highlighting](images/errorHighlighting.gif)

## Feature overview

* Syntax highlighting
* Autocomplete (for external packages and experimental for local project)
* Error highlighting
* Hover info
* Document Symbol provider (Go to definition, list and search symbols)
* Code Actions (Lightbulb on errors and warnings with fixes)
* Integration with Elm Package (Browse and install packages)
* REPL integration
* Integration with Elm Reactor
* Integration with Elm Make
* Custom Elm Snippets
* Code formatting with Elm-format
* Elm-analyse integration (experimental) with warnings and code actions

## Elm Installation

### Global Installation

Follow [this guide](https://guide.elm-lang.org/install.html).

### Project (Local) Installation

Run `npm install --save-dev elm`

Then, in `.vscode/settings.json`, add the following:

```
"elm.compiler": "./node_modules/.bin/elm",
"elm.makeCommand": "./node_modules/.bin/elm-make"
```

For Windows
```
"elm.compiler": ".\\node_modules\\.bin\\elm",
"elm.makeCommand": ".\\node_modules\\.bin\\elm-make"
```

## Feature details

### Syntax highlighting

Syntax highlighting is essential. The full language is supported. Fenced code blocks in markdown files are highlighted too. Can we improve the highlighting further? Please create an [issue](https://github.com/sbrink/vscode-elm/issues)!

### Error highlighting

We support error highlighting **on save**. If you check *Auto save* under File, you should get feedback immediately.

![Error highlighting](images/errorHighlighting.gif)

This is marked **experimental** because we still have to improve the project detection.

### Hover info - Function information

![Function info](images/functionInfo.gif)

You can hover over a function and get the signature and comment.

### Document and Workspace symbols

Use context menu "Go to definition" or F12. Alt+F12 to Peek

![Go to definition](images/gotoDefinition.gif)

Ctrl+Shift+O for document symbols and Ctrl+T for workspace symbols

![Search/browse document symbols and workspace symbols](images/searchDefinition.gif)

### Code Actions (Lightbulb on errors and warnings with fixes)

Tip - use Ctrl+. to invoke code action when the lightbulb is visible

![Code actions](images/codeActions.gif)

### Integration with Elm Package (Browse and install packages)

Ctrl+Shift+P - Elm browse packages

![Browse packages](images/browsePackages.gif)

Ctrl+Shift+P - Elm install package

![Install package](images/installPackage.gif)

### Local Project Intellisense (experimental)

vscode-elm will scan your projects to build intellisense results.
**Assumption** your files match the layout of elm-format

Example behaviors are:

* suggesting function names available in the current file or imported from your own modules
* suggesting module names you have imported in the current file
* suggesting union type or type alias names in function signatures
* suggesting possible values of a union type
* suggesting properties of a record. This requires that record is a parameter in the current function and the function has a type signature.

**important note regarding performance**

With `elm.userProjectImportStrategy` set to `"dynamicLookup"` (default), every hover or autocomplete action will trigger a scan of the first few lines of every file in your src directory (not the elm-stuff directory). This is done to establish an accurate list of module names but could be slow for exceptionally large projects.

In testing with the Elm SPA example project no slowdown was noticed, but if your project slows down try one of these alternate settings:

* `"semiDynamicLookup"` - the above directory scan still takes place but only once, then the results are stored in memory.  Subsequent hover or autocomplete results will be instantaneous but the list of available modules will only update if the window is reloaded.
* `"dotIsFolder"` - this assumes that any module name with a `.` in it is contained in a folder. For example Page.Home would attempt to look for Home.elm inside of the \Page\ folder.
* `"dotIsFilenameCharacter"` - this setting does not attempt to create a directory structure for Page.Home and instead looks for a file named \Page.Home.elm
* `"ignore"` - do not attempt to look in imported modules for intellisense


### REPL integration

![REPL](images/repl.gif)

Not sure about the output of a function? Test it from inside the editor.

Open the actions menu and use one of the following commands:

* Elm: REPL - Start
* Elm: REPL - Stop
* Elm: REPL - Send Line
* Elm: REPL - Send Selection
* Elm: REPL - Send File

### Reactor integration

![Reactor support](images/reactor.gif)

Reactor is the webserver which comes with Elm.

* Reactor allows recompiling on-the-fly.
* Reactor contains the [Time-traveling debugger](http://debug.elm-lang.org/).

We support starting / stopping from within the editor.

### Integration with Elm Make

* Elm: Make
* Elm: Make --warn
* Make user setting to choose a fixed file to make

### Snippets

We support snippets for the basic language features. To use them, press `Ctrl+Space` and start typing.
Or start with some characters and use `Ctrl+Space` for autocompletion.

Want to know more? Look at the [snippet definitions](snippets/elm.json)

### Formatting with Elm-format

[elm-format](https://github.com/avh4/elm-format) is supported via the editor's `Format Code` command. To format your code using `elm-format`, press `Shift+Alt+F` on Windows, `Shift+Option+F` on Mac, or `Ctrl+Shift+I` on Linux.

You can also configure `elm-format` to run on save by enabling `editor.formatOnSave` in your settings.

```
// settings.json
{
    "[elm]": {
        "editor.formatOnSave": true
    }
}
```

On Windows you may [need](https://github.com/Krzysztof-Cieslak/vscode-elm/issues/191#issuecomment-371465313) to add `editor.formatOnSaveTimeout`:

``` settings.json
"[elm]": {
    "editor.formatOnSave": true,
    "editor.formatOnSaveTimeout": 1500
},
```

### Elm-format - local installation
If you have elm-format installed locally add this line to '.vscode/settings.json'

```
"elm.formatCommand": "./node_modules/.bin/elm-format"
```

For Windows
```
"elm.formatCommand": ".\\node_modules\\.bin\\elm-format"
```


### Elm-analyse integration
[Elm-analyse](https://github.com/stil4m/elm-analyse) is a tool that allows you to analyse your Elm code, identify deficiencies and apply best practices.
The integration enables vscode to show any problems identified by elm-analyse as linting warnings in your code.

![Elm-analyse](images/elm-analyse.gif)

With a button to stop the elm-analyse process

![Elm-analyse stop](images/elm-analyse-stop.png)

To install elm-analyse
```
// bash/console
{
    npm install elm-analyse -g
}
```

Commands:

* To enable, run the command: Elm: `Start elm-analyse`
* To disable, run the comand: Elm: `Stop elm-analyse`
* When available - Code actions will show lightbulb to fix issue

When running, any issues will show up in the problems window (`Ctrl+Shift+M`), `F8` to cycle. Issues will also show up as green linting warnings in your code.

NOTE: *This is an early version*

* Elm-analyse and this integration is still under development. API will change.
* Running elm-analyse as a process inside vscode can cause performance issues. You can start elm-analyse as a separate process. When using the same port as specified in settings, the extension will detect the external process and use that instead of starting inside vscode.
* Mostly working, but if the elm-analyse process was not stopped: Linux: ps aux | grep elm-analyse $kill -9 {process id from ps} / On Windows - locate node elm-analyse process

Elm-analyse settings:

* `elm.analyseCommand` - Command to run when executing elm-analyse
* `elm.analysePort` - Port used by elm-analyse process
* `elm.analyseEnabled` - Enable or disable elm-analyse process on startup.

### Clean Build Artifacts

You can delete your `elm-stuff/build-artifacts` directly from vscode by using `Elm: Clean build artifacts` command.

## Acknowledgements

* Grammar file is taken and converted from [atom-elm](https://github.com/edubkendo/atom-elm).
* Initial snippets from [Elm.tmLanguage](https://github.com/deadfoxygrandpa/Elm.tmLanguage)

## How to contribute

*Imposter syndrome disclaimer*: I want your help. No really, I do.

There might be a little voice inside that tells you you're not ready; that you need to do one more tutorial, or learn another framework, or write a few more blog posts before you can help me with this project.

I assure you, that's not the case.

This project has some clear Contribution Guidelines and expectations that you can [read here](https://github.com/Krzysztof-Cieslak/vscode-elm/blob/master/CONTRIBUTING.md).

The contribution guidelines outline the process that you'll need to follow to get a patch merged. By making expectations and process explicit, I hope it will make it easier for you to contribute.

And you don't just have to write code. You can help out by writing documentation, tests, or even by giving feedback about this work. (And yes, that includes giving feedback about the contribution guidelines.)

Thank you for contributing!

## Contributing and copyright

The project is hosted on [GitHub](https://github.com/Krzysztof-Cieslak/vscode-elm/) where you can [report issues](https://github.com/Krzysztof-Cieslak/vscode-elm/issues), fork
the project and submit pull requests. Please read the [CONTRIBUTING](https://github.com/Krzysztof-Cieslak/vscode-elm/blob/master/CONTRIBUTING.md) document for more information.

The library is available under [MIT license](https://github.com/Krzysztof-Cieslak/vscode-elm/blob/master/LICENSE.md), which allows modification and redistribution for both commercial and non-commercial purposes.

Please note that this project is released with a [Contributor Code of Conduct](CODE_OF_CONDUCT.md). By participating in this project you agree to abide by its terms.

## Maintainer(s)

* Krzysztof Cieslak [@Krzysztof-Cieslak](https://github.com/Krzysztof-Cieslak)
* Dave Thomas [@7sharp9](https://github.com/7sharp9)
* Håkon Rossebø [@hakonrossebo](https://github.com/hakonrossebo)

## Past Maintainer(s)

* Sascha Brink [@sbrink](https://github.com/sbrink)
* Robert Jeppesen [@rojepp](https://github.com/rojepp)


CHANGELOG.md
### 0.26.0 - 16.12.2019

* Project deprecated. Use https://github.com/elm-tooling/elm-language-client-vscode instead

### 0.25.0 - 07.01.2019
* Fix issue with elm module parser using object spread in raw javascript (not transpiled). (by @joeandaverde)
* Fix issue with improperly invaliding old symbols in cache. (by @joeandaverde)
* Fix elm analyse path (by @Razzeee)
* The linter should also respect the module in "Make special path" (by @ccapndave)

### 0.24.0 - 26.11.2018
* Improved Symbol Search / Goto Definition (by @joeandaverde)
* Fix wrong missing compiler error message (by @andys8)
* Check isTestFile before normalizing path on windows (by @Krzysztof-Cieslak)

### 0.23.0 - 19.11.2018
* corrected a spelling error (by @WEBhicham)
* Replace inefficient regex to locate imports with simpler regex (by @joeandaverde)
* Add json schema support for Elm's config files (by @stephenreddek)
* Add support for running tests for Elm 0.19 (by @jackfranklin)

### 0.22.0 - 01.10.2018
* Fix a 0.18 elm-make configuration bug and support elm-format for 0.18 and 0.19 (by @hakonrossebo)
* Setting and logic to disable linting (by @hakonrossebo)
* Adds external package "oracle"-support for Elm 0.19 (by @r1sc)
* Handle breaking change in elm-format (by @hakonrossebo)

### 0.21.0 - 30.08.2018
* Set elm-make as default makeCommand for 0.18 and earlier (by @jackfranklin)
* Fix for Elm 0.19, find right project directory (by @norpan)
* Fix linting for errors for Elm 0.18 (by @jackfranklin)
* Changes REPL to use integrated terminal for interactivity (by @hakonrossebo)
* Make compiler errors a little easier to read (by @norpan)
* Separate example code for 0.18 and 0.19 (by @hakonrossebo)
* Fixes to omit Elm Oracle for 0.19 and read elm.json (by @hakonrossebo)
* Using terminal instead of output window to install packages for y/n (by @hakonrossebo)

### 0.20.0 - 24.08.2018
* Snippets: Let expression indentation (elm-format) (by @andys8)
* Add ability to configure elm-package locating. (by @jackfranklin)
* Syntax Polish (by @mdgriffith)
* Elm 0.19 update (by @hakonrossebo and @7sharp9)

### 0.19.0 - 27.07.2018
* Snippets: Add todo comment snippet (by @andys8)
* Refactoring diagnostic init for LS (by @lostintangent)
* Grammar updates (by @mdgriffith)
* Compilable Msg placeholder in HTML.program snippet (by @milesfrain)
* Info about editor.formatOnSaveTimeout on Windows (by @rofrol)

### 0.18.0 - 22.05.2018
* Make linting work again with elm-analyse 0.14.2 (by @Yarith)
* Handle formatSelection the same as formatDocument (by @milesfrain)

### 0.17.0 - 26.04.2018
* Limiting workspace symbols and linting to local files (by @lostintangent)
* Fixed issue with elm-make and spaces in paths (by @Arbyy)
* Fix error message not clearing when formatting with elm-format (by @adam77)
* Make folding not eat up empty lines (by @szabba)

### 0.16.0 - 11.02.2018
*  Jump to symbol based on imports. Jump to symbol that has been fully qualified such as `MyModule.myFunction` (by @joeandaverde)
* Fix cursor moving when formatting with elm-format (by @adam77)
* Support elm code syntax highlighting in markdown files (by @kachkaev)
* Snippets improvement (by @andys8)
* Fix `elm.makeCommand` sample path for Windows (by @lukaszgasior)

### 0.15.1 - 03.10.2017
* added prettier to the project (by @littleStudent)
* Fix package installation  (by @wolverian)
* save file only once formatting is finished (by @gyzerok)
* Snippets for elm-test test and describe blocks (by @andys8 )

### 0.15.0 - 10.08.2017
* Add integration with `Elm-analyse` (by @hakonrossebo)
* Use TsLint on `vscode-elm` code base (by @littleStudent)

### 0.14.1 - 21.06.2017
* Added the ability to include function/type/type alias comments in intellisense (by @andrewsdev)
* Fixed bugs in parsing elm-oracle hover results and bugs in userProject hover results (by @andrewsdev)

### 0.14.0 - 15.06.2017
* Fix defaultModel snippet, Add Html.program snippet (by @iocube)
* Bugfix: Error with one-line suggested annotation (by @hakonrossebo)
* Handle block comments (by @hakonrossebo)
* Improvements/bug fixes for user project intellisense (by @andrewsdev):
  - Bug fix: writing type signature before writing the function definition behavior
  - Bug fix: `/as/` regex for the list of imports behavior
  - Bug fix: name of a user's type or type alias were not included in intellisense results
  - Feature: Primitive types are now included in autocomplete when writing a function definition
  - Feature: limit on user intellisense 'comment' size now better indicates it has hit a settings limit
  - Feature: Can now set `elm.userProjectMaxCommentSize` to 0 to bypass the limit on user intellisense 'comment' size
  - Feature: Working Intellisense (hover and autocomplete) for user modules with a '.'. Custom configuration to control this feature.

### 0.13.0 - 08.06.2017
* Workspace symbols implementation (by @hakonrossebo)
* Elm-reactor subdir traversal (by @hakonrossebo)
* Command to stop Elm REPL (by @hakonrossebo)
* Elm-make uses path correctly. Also introduce special file config (by @hakonrossebo)

### 0.12.0 - 03.06.2017
* add auto-closing quotes to config (by @kojuro-kun)
* Changes snippets to 4 space indentation (by @hakonrossebo)
* Fixes double module name in autocomplete (by @hakonrossebo)
* Autocomplete handles aliased module names (by @hakonrossebo)
* Elm-format error as StatusBar instead of ErrorMessage (by @hakonrossebo)
* Initial implementation of CodeActions (function type annotations, misspelled variable names, types and patterns) (by @hakonrossebo)
* Implements go to definition for document (by @hakonrossebo)

### 0.11.2 - 10.05.2017
* Improved Linter so that issues are associated with the correct files (by @danheuck )
* Fix for elm-oracle being fed entire elm file (by @dannyob)

### 0.11.1 - 10.05.2017
* Fix keyboard shortcuts

### 0.11.0 - 08.05.2017
* Add `Elm: Browse package` command
* Fix paths concatenation issue for non-Windows machines

### 0.10.0 - 05.05.2017


src/elmLinter.ts
import * as cp from 'child_process';
import * as readline from 'readline';
import * as path from 'path';
import * as utils from './elmUtils';
import * as vscode from 'vscode';
import * as elmTest from './elmTest';

import { ElmAnalyse } from './elmAnalyse';

export interface IElmIssueRegion {
  start: { line: number; column: number };
  end: { line: number; column: number };
}

export interface IElmIssue {
  tag: string;
  overview: string;
  subregion: string;
  details: string;
  region: IElmIssueRegion;
  type: string;
  file: string;
}

function severityStringToDiagnosticSeverity(
  severity: string,
): vscode.DiagnosticSeverity {
  switch (severity) {
    case 'error':
      return vscode.DiagnosticSeverity.Error;
    case 'warning':
      return vscode.DiagnosticSeverity.Warning;
    default:
      return vscode.DiagnosticSeverity.Error;
  }
}

function elmMakeIssueToDiagnostic(issue: IElmIssue): vscode.Diagnostic {
  let lineRange: vscode.Range = new vscode.Range(
    issue.region.start.line - 1,
    issue.region.start.column - 1,
    issue.region.end.line - 1,
    issue.region.end.column - 1,
  );
  return new vscode.Diagnostic(
    lineRange,
    issue.overview + ' - ' + issue.details.replace(/\[\d+m/g, ''),
    severityStringToDiagnosticSeverity(issue.type),
  );
}

function parseErrorsElm019(line, file) {
  const returnLines = [];
  const errorObject = JSON.parse(line);

  if (errorObject.type === 'compile-errors') {
    errorObject.errors.forEach(error => {
      const problems = error.problems.map(problem => ({
        tag: 'error',
        overview: problem.title,
        subregion: '',
        details: problem.message
          .map(
            message =>
              typeof message === 'string'
                ? message
                : '#' + message.string + '#',
          )
          .join(''),
        region: problem.region,
        type: 'error',
        file,
      }));

      returnLines.push(...problems);
    });
  } else if (errorObject.type === 'error') {
    const problem = {
      tag: 'error',
      overview: errorObject.title,
      subregion: '',
      details: errorObject.message
        .map(
          message => (typeof message === 'string' ? message : message.string),
        )
        .join(''),
      region: {
        start: {
          line: 1,
          column: 1,
        },
        end: {
          line: 1,
          column: 1,
        },
      },
      type: 'error',
      file: errorObject.path,
    };

    returnLines.push(problem);
  }

  return returnLines;
}

function parseErrorsElm018(line) {
  if (line.startsWith('Successfully generated')) {
    // ignore compiler successes
    return [];
  }
  // elm make returns an array of issues
  return <IElmIssue[]>JSON.parse(line);
}

function checkForErrors(fullFilename): Promise<IElmIssue[]> {
  return new Promise((resolve, reject) => {
    const config: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration(
      'elm',
    );
    const make018Command: string = <string>config.get('makeCommand');
    const compiler: string = <string>config.get('compiler');
    const elmTestCompiler: string = <string>config.get('elmTestCompiler');
    const [cwd, elmVersion] = utils.detectProjectRootAndElmVersion(
      fullFilename,
      vscode.workspace.rootPath,
    );
    const specialFile: string = <string>config.get('makeSpecialFile');
    const isTestFile = elmTest.fileIsTestFile(fullFilename);
    let make;

    let filename = utils.isElm019(elmVersion)
      ? path.relative(cwd, fullFilename)
      : fullFilename;

    if (specialFile.length > 0) {
      filename = path.resolve(cwd, specialFile);
    }
    if (utils.isWindows) {
      filename = '"' + filename + '"';
    }

    const args018 = [filename, '--report', 'json', '--output', '/dev/null'];
    const args019 = [
      'make',
      filename,
      '--report',
      'json',
      '--output',
      '/dev/null',
    ];
    const args = utils.isElm019(elmVersion) ? args019 : args018;
    const makeCommand = utils.isElm019(elmVersion)
      ? isTestFile
        ? elmTestCompiler
        : compiler
      : make018Command;

    if (utils.isWindows) {
      make = cp.exec(makeCommand + ' ' + args.join(' '), { cwd: cwd });
    } else {
      make = cp.spawn(makeCommand, args, { cwd: cwd });
    }
    // output is actually optional
    // (fixed in https://github.com/Microsoft/vscode/commit/b4917afe9bdee0e9e67f4094e764f6a72a997c70,
    // but unreleased at this time)
    const errorLinesFromElmMake = readline.createInterface({
      // elm 0.19 uses stderr, elm 0.18 uses stdout
      input: utils.isElm019(elmVersion) ? make.stderr : make.stdout,
      output: undefined,
    });

    const lines = [];

    const elm018stderr: Buffer[] = [];

    errorLinesFromElmMake.on('line', line => {
      if (utils.isElm019(elmVersion)) {
        const newLines = parseErrorsElm019(line, fullFilename);
        newLines.forEach(l => lines.push(l));
      } else {
        const newLines = parseErrorsElm018(line);
        newLines.forEach(l => lines.push(l));
      }
    });

    if (utils.isElm019(elmVersion) === false) {
      // we listen to stderr for Elm 0.18
      // as this is where a whole file issue would go
      make.stderr.on('data', (data: Buffer) => {
        if (data) {
          elm018stderr.push(data);
        }
      });
    }

    make.on('error', err => {
      errorLinesFromElmMake.close();
      if (err && err.code === 'ENOENT') {
        vscode.window.showInformationMessage(
          `The elm compiler is not available (${makeCommand}). Install Elm from https://elm-lang.org.`,
        );
        resolve([]);
      } else {
        reject(err);
      }
    });

    make.on('close', (code, signal) => {
      errorLinesFromElmMake.close();

      if (elm018stderr.length) {
        let errorResult: IElmIssue = {
          tag: 'error',
          overview: '',
          subregion: '',
          details: elm018stderr.join(''),
          region: {
            start: {
              line: 1,
              column: 1,
            },
            end: {
              line: 1,
              column: 1,
            },
          },
          type: 'error',
          file: filename,
        };
        resolve([errorResult]);
      } else {
        resolve(lines);
      }
    });
  });
}

let compileErrors: vscode.DiagnosticCollection;

export function runLinter(
  document: vscode.TextDocument,
  elmAnalyse: ElmAnalyse,
): void {
  if (document.languageId !== 'elm' || document.uri.scheme !== 'file') {
    return;
  }

  let uri: vscode.Uri = document.uri;

  if (!compileErrors) {
    compileErrors = vscode.languages.createDiagnosticCollection('elm');
  } else {
    compileErrors.clear();
  }

  checkForErrors(uri.fsPath)
    .then((compilerErrors: IElmIssue[]) => {
      const cwd: string =
        utils.detectProjectRoot(uri.fsPath) || vscode.workspace.rootPath;
      let splitCompilerErrors: Map<string, IElmIssue[]> = new Map();

      compilerErrors.forEach((issue: IElmIssue) => {
        // If provided path is relative, make it absolute
        if (issue.file.startsWith('.')) {
          issue.file = cwd + issue.file.slice(1);
        }
        if (splitCompilerErrors.has(issue.file)) {
          splitCompilerErrors.get(issue.file).push(issue);
        } else {
          splitCompilerErrors.set(issue.file, [issue]);
        }
      });
      // Turn split arrays into diagnostics and associate them with correct files in VS
      splitCompilerErrors.forEach((issue: IElmIssue[], issuePath: string) => {
        compileErrors.set(
          vscode.Uri.file(issuePath),
          issue.map(error => elmMakeIssueToDiagnostic(error)),
        );
      });
    })
    .catch(error => {
      compileErrors.set(document.uri, []);
    });

  if (elmAnalyse.elmAnalyseIssues.length > 0) {
    let splitCompilerErrors: Map<string, IElmIssue[]> = new Map();
    elmAnalyse.elmAnalyseIssues.forEach((issue: IElmIssue) => {
      if (splitCompilerErrors.has(issue.file)) {
        splitCompilerErrors.get(issue.file).push(issue);
      } else {
        splitCompilerErrors.set(issue.file, [issue]);
      }
      splitCompilerErrors.forEach(
        (analyserIssue: IElmIssue[], issuePath: string) => {
          compileErrors.set(
            vscode.Uri.file(issuePath),
            analyserIssue.map(error => elmMakeIssueToDiagnostic(error)),
          );
        },
      );
    });
  }
}
